### PR TITLE
Pista_Marcel_SM_TM_5317_I_want_to_be_able_to_assign_a_model_to_Client…

### PR DESCRIPTION
**Pull request by pistamarcel**
Pista_Marcel_SM_TM_5317_I_want_to_be_able_to_assign_a_model_to_Clients_and_Endusers first draft

### CODE REVIEW

File: src/main/java/pages/AssignAModelToClientsAndEndUsersPage.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
+package pages;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import java.util.List;
+
+public class AssignAModelToClientsAndEndUsersPage extends BasePage {
```
Comment (95teo) [inline src/main/java/pages/AssignAModelToClientsAndEndUsersPage.java:9]: Can you please rename the class to AssignModelPage? Thanks!

File: src/main/java/pages/AssignAModelToClientsAndEndUsersPage.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    private WebElement closeButton;
+
+
+    public void clickClientPage() throws InterruptedException {
+        Thread.sleep(1000);
+        elementControl.clickElement(clientsBtnPage);
+    }
+
+    public void clickEndUserPage() throws InterruptedException {
+        Thread.sleep(1000);
```
Comment (95teo) [inline src/main/java/pages/AssignAModelToClientsAndEndUsersPage.java:45]: Can you please replace Thread.sleep() occurrences from this class with some new/existing methods from WaitUtils that will use FluentWait? For example, waitForElementToBePresent/Visible, waitForElementToBeInteractable. Let me know if you need more guidance. Thanks!

File: src/test/java/assignAModelTo/AssignAModelToClientsAndEndUsersTest.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+public class AssignAModelToClientsAndEndUsersTest extends Base {
+
+
+    @Test(description = "Check and uncheck a model for Clients and End Users as admin", dataProvider = "dataProvider")
+    public void assignAModelFromAnAdminAccount(boolean client, boolean uncheck) throws InterruptedException {
+        loginPage.loginWithCredentials("capnv12@gmail.com", "Capnv12");
+        if (client) {
+            if (uncheck) {
+                assignAModelToClientsAndEndUsersPage.unCheckClientWithSaving();
+                Thread.sleep(1000);
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelToClientsAndEndUsersTest.java:19]: Can we replace these Thread.sleep() occurrences with methods that use FluentWait behind? Thanks!

File: src/test/java/assignAModelTo/AssignAModelToClientsAndEndUsersTest.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+import static org.testng.Assert.*;
+
+public class AssignAModelToClientsAndEndUsersTest extends Base {
+
+
+    @Test(description = "Check and uncheck a model for Clients and End Users as admin", dataProvider = "dataProvider")
+    public void assignAModelFromAnAdminAccount(boolean client, boolean uncheck) throws InterruptedException {
+        loginPage.loginWithCredentials("capnv12@gmail.com", "Capnv12");
+        if (client) {
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelToClientsAndEndUsersTest.java:16]: It's not recommended to have this logic at test level. Let's make separate scenarios for each of the if-else code branches. You can use some dataProviders for username/password maybe in order not to hardcode them in every test case. Thanks!

**Review by 95teo — CHANGES_REQUESTED** [review]

File: src/test/java/assignAModelTo/AssignAModelToClientsAndEndUsersTest.java (hunk 4 above)
Comment (95teo) [reply src/test/java/assignAModelTo/AssignAModelToClientsAndEndUsersTest.java:16]: To be honest I don't like what I hear ... that means that your tests depend on each other and are not independent ... Can we have different models for each test case? Is it duable? Thanks! ;)

File: src/main/java/pages/AssignModelPage.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+public class AssignModelPage extends BasePage {
+
+    public AssignModelPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public static final String CHECK_ELEMENT_CLIENT_CSS = "div[class='w-full flex justify-between items-center my-2 p-2 relative cursor-pointer bg-primary bg-opacity-20']";
+    public static final String UNCHECK_ELEMENT_CLIENT_CSS = "div[class='w-full flex justify-between items-center my-2 p-2 relative cursor-pointer false']";
+    public static final String ASSIGN_MODEL_BUTTON_LIST_CSS = "div[style='flex-direction: column; display: flex; gap: 10px;']";
+    @FindBy(css = "li.rounded-lg.mb-2 a[href='/clients']")
```
Comment (95teo) [inline src/main/java/pages/AssignModelPage.java:20]: Please leave a blank line above in order to separate a bit @FindBy section from String declaration section. Thanks!

File: src/main/java/pages/AssignModelPage.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    public List<WebElement> getAssignModelButtonList() {
+        return assignModelButtonList;
+    }
+
+    public WebElement getCloseButton() {
+        return closeButton;
+    }
+
+    public void clickClientPage() {
+        WaitUtils waitUtils = new WaitUtils(driver);
```
Comment (95teo) [inline src/main/java/pages/AssignModelPage.java:54]: Try to put this line with initialization WaitUtils waitUtils = new WaitUtils(driver); as a member in the class and not in every method that it's used.

File: src/main/java/pages/AssignModelPage.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public void checkEndUserWithoutSaving() {
+        WaitUtils waitUtils = new WaitUtils(driver);
+        waitUtils.waitForVisibilityOfAllElements(By.cssSelector(CHECK_ELEMENT_CLIENT_CSS));
+        elementControl.clickElement(unCheckedModelsList.get(0));
+        elementControl.clickElement(getCloseButton());
+        waitUtils.waitForElementToBeMoreThan(By.cssSelector(ASSIGN_MODEL_BUTTON_LIST_CSS), 5);
+        clickAssignModelTo();
+
```
Comment (95teo) [inline src/main/java/pages/AssignModelPage.java:130]: Remove this line. Thanks!

File: src/main/java/utils/WaitUtils.java [hunk 8]
```diff
@@ -18,5 +18,11 @@ public WaitUtils(WebDriver driver) {
     public void waitForElementToBeMoreThan(By locator, Integer number) {
         fluentWait.until(ExpectedConditions.numberOfElementsToBeMoreThan(locator, number));
     }
+    public void waitForElementToBeLessThan(By locator, Integer number) {
```
Comment (95teo) [inline src/main/java/utils/WaitUtils.java:21]: Please leave a blank line between methods. Thanks!

File: src/test/java/assignAModelTo/AssignAModelFromAdmin.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+import static org.testng.Assert.*;
+import static pages.AssignModelPage.*;
+
+public class AssignAModelFromAdmin extends Base {
+
+    @Test(description = "Check a model for Clients as admin", dataProvider = "dataProvider")
+    public void assignAModelToClient(String username, String password) {
+        loginPage.loginWithCredentials(username, password);
+        WaitUtils waitUtils = new WaitUtils(driver);
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelFromAdmin.java:18]: Put this line with initialization of WaitUtils inside Base test class and remove it from this class. Thanks!

File: src/test/java/assignAModelTo/AssignAModelFromAdmin.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    @Test(description = "Check a model for End Users as admin", dataProvider = "dataProvider")
+    public void assignAModelToEndUser(String username, String password)  {
+        loginPage.loginWithCredentials(username, password);
+        WaitUtils waitUtils = new WaitUtils(driver);
+        assignModelPage.clickEndUserPage();
+        waitUtils.waitForElementToBeMoreThan(By.cssSelector(ASSIGN_MODEL_BUTTON_LIST_CSS), 4);
+        assignModelPage.clickAssignModelTo();
+        ;
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelFromAdmin.java:106]: Do you miss a line here? If not, please remove line with ';' Thanks!

File: src/test/java/assignAModelTo/AssignAModelFromAdmin.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+            this.logger = Logger.getLogger(this.getClass().getName());
+            logger.info("Checked models list size before test for client:************* " + beforeTest);
+            logger.info("Checked models list size after test for client:************* " + afterTest);
+            assertEquals(afterTest, beforeTest, "List of checked models has  been modified after test  even tough we didn't saved our option!BUG!");
+        } else {
+            this.logger = Logger.getLogger(this.getClass().getName());
+            logger.info("There are no models for us to uncheck on this client");
+        }
+    }
+
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelFromAdmin.java:184]: Remove on line from here, please!

File: src/test/java/assignAModelTo/AssignAModelFromAdmin.java [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+    @DataProvider
+    public Object[][] dataProvider() {
+        return new Object[][]{
+                {"capnv12@gmail.com", "Capnv12"},};
+    }
+}
+
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelFromAdmin.java:192]: Leave only one blank line here. Thanks!

File: src/test/java/assignAModelTo/AssignAModelFromSupport.java [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+import static org.testng.Assert.assertFalse;
+import static pages.AssignModelPage.*;
+import static pages.AssignModelPage.UNCHECK_ELEMENT_CLIENT_CSS;
+
+public class AssignAModelFromSupport extends Base {
+
+    @Test(description = "Check a model for Clients as Support", dataProvider = "dataProvider")
+    public void assignAModelToClient(String username, String password) {
+        loginPage.loginWithCredentials(username, password);
+        WaitUtils waitUtils = new WaitUtils(driver);
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelFromSupport.java:20]: Put this line with initialization of WaitUtils inside Base test class and remove it from this class. Thanks!

File: src/test/java/assignAModelTo/AssignAModelFromSupport.java [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+
+    @DataProvider
+    public Object[][] dataProvider() {
+        return new Object[][]{
+                {"marcel.pista96@e-uvt.ro", "dvsbrho5"},};
+    }
+}
+
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelFromSupport.java:195]: Leave only one blank line here. Thanks!

File: src/test/java/assignAModelTo/AssignAModelFromSupport.java [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+            logger.info("Checked models list size before test for client:************* " + beforeTest);
+            logger.info("Checked models list size after test for client:************* " + afterTest);
+            assertEquals(afterTest, beforeTest,
+                    "List of checked models has  been modified after test  even tough we didn't saved our option!BUG!");
+        } else {
+            this.logger = Logger.getLogger(this.getClass().getName());
+            logger.info("There are no models for us to uncheck on this client");
+        }
+    }
+
```
Comment (95teo) [inline src/test/java/assignAModelTo/AssignAModelFromSupport.java:187]: Leave only one blank line here. Thanks!

**Review by 95teo — CHANGES_REQUESTED** [review]

**Review by 95teo — APPROVED** [review]

**Comment by 95teo** [comment]
Please fix conflicts first. Thanks!

**Review by 95teo — APPROVED** [review]

File: src/test/java/assignAModelTo/Base.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+import utils.ConfigUtils;
+import utils.ReportUtils;
+import utils.ScreenshotUtils;
+import utils.WaitUtils;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Properties;
+import java.util.logging.Logger;
+
```
Comment (stanciuvlad99) [inline src/test/java/assignAModelTo/Base.java:21]: Delete the blank space. Thank you!

**Review by stanciuvlad99 — APPROVED** [review]

**Pull request merged**